datumaro/plugins/synthetic_data/image_generator.py

```diff
@@ -0,0 +1,160 @@
+                rng, param, self._iterations, self._height, self._width, draw_point=False, weight=w
+            )
+            color_image = colorize(image, net)
+            aug_image = augment(rng, color_image, synthetic_background)
+            cv.imwrite(osp.join(self._output_dir, "{:06d}.png".format(i)), aug_image)
+
+    def _generate_image(
+        self,
+        rng: Random,
+        params: np.ndarray,
+        iterations: int,
+        height: int,
+        width: int,
+        draw_point: bool = True,
+        weight: Optional[np.ndarray] = None,
+    ) -> np.ndarray:
+        ifs_function = IFSFunction(rng, prev_x=0.0, prev_y=0.0)
+        for param in params:
+            ifs_function.add_param(param[:6], param[6], weight)
+        ifs_function.calculate(iterations)
+        img = ifs_function.draw(height, width, draw_point)
+        return img
+
+    def _generate_category(self, rng: Random, base_h: int = 512, base_w: int = 512) -> np.ndarray:
+        pixels = -1
+        while pixels < self._threshold:
+            param_size = rng.randint(2, 7)
+            params = np.zeros((param_size, 7), dtype=np.float32)
+
+            sum_proba = 1e-5
+            for i in range(param_size):
+                a, b, c, d, e, f = [rng.uniform(-1.0, 1.0) for _ in range(6)]
+                prob = abs(a * d - b * c)
+                sum_proba += prob
+                params[i] = a, b, c, d, e, f, prob
+            params[:, 6] /= sum_proba
+
+            fractal_img = self._generate_image(rng, params, self._num_of_points, base_h, base_w)
+            pixels = np.count_nonzero(fractal_img) / (base_h * base_w)
+        return params
+
+    def _initialize_params(self) -> None:
+        default_img_size = 362 * 362
+        points_coeff = max(1, int(np.round(self._height * self._width / default_img_size)))
+        self._num_of_points = 100000 * points_coeff
+
+        instances_categories = np.ceil(self._count / self._weights.shape[0])
+        self._instances = np.ceil(np.sqrt(instances_categories)).astype(np.int)
+        self._categories = np.floor(np.sqrt(instances_categories)).astype(np.int)
+
+        if self._count < self._weights.shape[0]:
+            self._weights = self._weights[: self._count, :]
+
+    @staticmethod
+    def _create_weights(num_params):
+        # weights from https://openaccess.thecvf.com/content/ACCV2020/papers/Kataoka_Pre-training_without_Natural_Images_ACCV_2020_paper.pdf
+        BASE_WEIGHTS = np.ones((num_params,))
+        WEIGHT_INTERVAL = 0.4
+        INTERVAL_MULTIPLIERS = (-2, -1, 1, 2)
+        weight_vectors = [BASE_WEIGHTS]
+
+        for weight_index in range(num_params):
+            for multiplier in INTERVAL_MULTIPLIERS:
+                modified_weights = BASE_WEIGHTS.copy()
+                modified_weights[weight_index] += multiplier * WEIGHT_INTERVAL
+                weight_vectors.append(modified_weights)
+        weights = np.array(weight_vectors).reshape(-1, 6)
```

Comment:
I see this 6 in several places, it looks like a good candidate for a class constant.

Reply:
Changed to IFSFunction.NUM_PARAMS